### ISSUE TITLE
tf.numpy_function with lambda function is much slower starting from TF 1.15

### ISSUE THREAD

**Issue by mspinaci** [tensorflow/tensorflow#37488]
**System information** 
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Linux Ubuntu 18.04 (our machine), and also COLAB.**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: **No**
- TensorFlow installed from (source or binary): **binary**
- TensorFlow version (use command below): **1.14 vs 1.15.2 vs 2.1.0**
- Python version: **3.6.9**
- Bazel version (if compiling from source): **nope**
- GCC/Compiler version (if compiling from source): **nope**
- CUDA/cuDNN version: **10.1 / 7.6.2**
- GPU model and memory: **V100 (our machine) and T4 in COLAB. But the issue is on the CPU side**

**Describe the current behavior**
When using `tf.numpy_function` (or `tf.compat.v1.py_func`) in combination with a lambda function (in order to pass some fixed, non-tensor parameters to it), performance decreased massively from 1.14 to 1.15 onwards. (depending on the problem and the hardware, in my tests between 2x and 8x; the provided code is about 2x slower on a Colab T4 instance, and 4x slower on a V100 machine; some other much more involved code got slowed down up to 8x; but that's mainly because with slower GPUs, the bottleneck may lie elsewhere; I haven't done specific tests using one line only)

**Describe the expected behavior**
The invoked function being quite literally the same, the time should be also virtually identical. And more generally, code shouldn't get slower with a new code version...

**Standalone code to reproduce the issue** 
Provide a reproducible test case that is the bare minimum necessary to generate the problem. If possible, please share a link to Colab/Jupyter/any notebook.

Link to Colab: https://colab.research.google.com/drive/1MZVOBlAY7piqgeKwGzA8goeU3xtPy5gp

The code above runs a full "training" pipeline using a keras model, estimators framework, and data generation via numpy_function. All data is created randomly, so the model won't learn anything of course. The main difference is between the line:

`tf.numpy_function(a_slow_numpy_function, [tensor, [x1, y1, x2, y2]], tf.float32)`

and the line

`tf.numpy_function(lambda ex: a_slow_numpy_function(ex, [x1, y1, x2, y2]), [tensor], tf.float32`

Here, tensor is... a tensor, while x1, y1, x2, y2 are integers. Therefore, both lines are valid, in the first one the list is implicitly converted to a tensor (and then to a numpy array), in the second one a temporary function has those parameters as fixed.

Until TF 1.14 both lines produced the same runtime, from TF 1.15 on the second one gets much slower (the first one stays the same).

However, unfortunately I wasn't able to obtain reproducible results by only invoking the line above in a standalone session.

**Other info / logs** 

Logs on the Colab T4 can be seen from the link (for all TF versions 1.14, 1.15, 2.1). The following logs were obtained on our V100 machine:

**TF 1.14, no lambda:**
INFO:tensorflow:loss = 0.2136189, step = 0
INFO:tensorflow:global_step/sec: 32.7136
INFO:tensorflow:loss = 0.03677105, step = 100 (3.057 sec)
INFO:tensorflow:global_step/sec: 38.7334
INFO:tensorflow:loss = 0.038587682, step = 200 (2.581 sec)
INFO:tensorflow:global_step/sec: 38.3866
INFO:tensorflow:loss = 0.037003815, step = 300 (2.605 sec)
INFO:tensorflow:global_step/sec: 38.1023
INFO:tensorflow:loss = 0.037481233, step = 400 (2.625 sec)
INFO:tensorflow:global_step/sec: 27.7657
INFO:tensorflow:loss = 0.03703531, step = 500 (3.602 sec)
INFO:tensorflow:global_step/sec: 38.0632
INFO:tensorflow:loss = 0.03458761, step = 600 (2.627 sec)
INFO:tensorflow:global_step/sec: 36.4805
INFO:tensorflow:loss = 0.03637625, step = 700 (2.741 sec)
INFO:tensorflow:global_step/sec: 35.7584
INFO:tensorflow:loss = 0.036259696, step = 800 (2.797 sec)
INFO:tensorflow:global_step/sec: 36.7052
INFO:tensorflow:loss = 0.036455855, step = 900 (2.725 sec)

**TF 1.14, with lambda:**
INFO:tensorflow:loss = 0.23772891, step = 0
INFO:tensorflow:global_step/sec: 32.9305
INFO:tensorflow:loss = 0.03767074, step = 100 (3.037 sec)
INFO:tensorflow:global_step/sec: 37.3223
INFO:tensorflow:loss = 0.03685203, step = 200 (2.679 sec)
INFO:tensorflow:global_step/sec: 36.7882
INFO:tensorflow:loss = 0.038278855, step = 300 (2.718 sec)
INFO:tensorflow:global_step/sec: 37.7637
INFO:tensorflow:loss = 0.03842716, step = 400 (2.648 sec)
INFO:tensorflow:global_step/sec: 27.7136
INFO:tensorflow:loss = 0.036828667, step = 500 (3.608 sec)
INFO:tensorflow:global_step/sec: 35.5345
INFO:tensorflow:loss = 0.03633987, step = 600 (2.814 sec)
INFO:tensorflow:global_step/sec: 37.1818
INFO:tensorflow:loss = 0.037201233, step = 700 (2.690 sec)
INFO:tensorflow:global_step/sec: 37.2266
INFO:tensorflow:loss = 0.036516882, step = 800 (2.686 sec)
INFO:tensorflow:global_step/sec: 38.3496
INFO:tensorflow:loss = 0.039767925, step = 900 (2.608 sec)

**TF 1.15, no lambda:**
INFO:tensorflow:loss = 0.25934136, step = 0
INFO:tensorflow:global_step/sec: 34.4488
INFO:tensorflow:loss = 0.037474792, step = 100 (2.903 sec)
INFO:tensorflow:global_step/sec: 37.6249
INFO:tensorflow:loss = 0.037868354, step = 200 (2.658 sec)
INFO:tensorflow:global_step/sec: 38.2594
INFO:tensorflow:loss = 0.0373457, step = 300 (2.614 sec)
INFO:tensorflow:global_step/sec: 39.2323
INFO:tensorflow:loss = 0.036336176, step = 400 (2.549 sec)
INFO:tensorflow:global_step/sec: 26.6403
INFO:tensorflow:loss = 0.035495166, step = 500 (3.754 sec)
INFO:tensorflow:global_step/sec: 37.6238
INFO:tensorflow:loss = 0.038571935, step = 600 (2.658 sec)
INFO:tensorflow:global_step/sec: 37.4171
INFO:tensorflow:loss = 0.03556432, step = 700 (2.672 sec)
INFO:tensorflow:global_step/sec: 35.1526
INFO:tensorflow:loss = 0.035202924, step = 800 (2.845 sec)
INFO:tensorflow:global_step/sec: 35.4907
INFO:tensorflow:loss = 0.03740203, step = 900 (2.818 sec)
INFO:tensorflow:Loss for final step: 0.03607688.

**TF 1.15, with lambda:**
INFO:tensorflow:loss = 0.25370103, step = 0
INFO:tensorflow:global_step/sec: 8.59509
INFO:tensorflow:loss = 0.037242886, step = 100 (11.635 sec)
INFO:tensorflow:global_step/sec: 8.02817
INFO:tensorflow:loss = 0.03554675, step = 200 (12.456 sec)
INFO:tensorflow:global_step/sec: 12.4317
INFO:tensorflow:loss = 0.03646993, step = 300 (8.044 sec)
INFO:tensorflow:global_step/sec: 8.62791
INFO:tensorflow:loss = 0.035594724, step = 400 (11.590 sec)
INFO:tensorflow:global_step/sec: 8.80279
INFO:tensorflow:loss = 0.036044836, step = 500 (11.360 sec)
INFO:tensorflow:global_step/sec: 9.01617
INFO:tensorflow:loss = 0.036771618, step = 600 (11.091 sec)
INFO:tensorflow:global_step/sec: 9.50426
INFO:tensorflow:loss = 0.038547777, step = 700 (10.522 sec)
INFO:tensorflow:global_step/sec: 9.18885
INFO:tensorflow:loss = 0.03686235, step = 800 (10.883 sec)

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please find the Gist [here](https://colab.sandbox.google.com/gist/amahendrakar/54472b5156b814d46865161a8f93daea/37488.ipynb). Thanks!

**Comment by ymodak** [comment]
I tested the code snippet with TF 2.5 and I see identical time elapsed for both versions with and without lambda.
